src/CacheWithResilience.php

```diff
@@ -0,0 +1,165 @@
+            // when we want the item to be consider as expired, the second one is the real TTL that we tell redis.
+            // The second one is only mean to be used in case of issues like DB down and only requires to
+            // set $returnStaleValue to true
+            if (false === $returnStaleValue && new DateTime() > $value['ttl']) {
+                return new CacheItem();
+            }
+        }
+
+        // in setItem() value gets wrapped into an array with the ttl. set the item to its original value
+        $val = $item->get();
+        $item->set($val['value']);
+
+        return $item;
+    }
+
+    /**
+     * @param CacheItemInterface $item
+     * @param mixed $value
+     * @param int|string|DateTimeInterface $ttl
+     *   TTL in seconds, or a constant from CacheInterface, or a DateTime to expire at
+     * @return bool
+     */
+    public function setItem(CacheItemInterface $item, $value, $ttl): bool
+    {
+        // this condition calculates the DateTime when the element should be consider as expire
+        if ($ttl instanceof DateTimeInterface) {
+            $expiresAt = $ttl;
+        } elseif (\in_array($ttl, [0, -1, CacheInterface::NONE, CacheInterface::INDEFINITE, CacheInterface::SHORT], true)) {
+            // this case handles a list of exception that shouldn't be cached for a different length
+            $ttl = $this->calculateTtl($ttl);
+            $item->expiresAfter($ttl);
+            // getItem() expects the item to be wrapped as [value, ttl]
+            $item->set(['value' => $value, 'ttl' => $ttl]);
+            return $this->cachePool->save($item);
+        } else {
+            $ttl = $this->calculateTtl($ttl);
+            $expiresAt = new DateTime();
+            $expiresAt = $expiresAt->add(new DateInterval('PT' . $ttl . 'S'));
+        }
+
+        // save the ttl within the stored item so we can know if a returned key is expired.
+        $item->set(['value' => $value, 'ttl' => $expiresAt]);
+
+        // instead of setting the item to expire at $expiresAt we use "RESILIENCE_TTL"
+        // this way in case of error we can return the stale value
+        $item->expiresAfter(self::RESILIENCE_TTL);
+
+        return $this->cachePool->save($item);
+    }
+
+    /**
+     * IF CALLABLE RETURNS SOMETHING THAT EVALUATES TO EMPTY THE RESULT WILL NOT BE CACHED UNLESS $nullTtl IS SET
+     * TO A VALUE DIFFERENT FROM CacheInterface::NONE
+     *
+     * @param string $key
+     * @param int|string $ttl
+     *   TTL in seconds, or a constant from CacheInterface
+     * @param callable $function
+     * @param array $arguments
+     * @return mixed
+     */
+    public function getOrSet(string $key, $ttl, callable $function, array $arguments = [], $nullTtl = CacheInterface::NONE)
+    {
+        $cacheItem = $this->getItem($key);
+
+        if ($cacheItem->isHit()) {
+            return $cacheItem->get();
+        }
+
+        try {
+            $result = $function(...$arguments);
```

Comment:
1. how about naming `$function` to reflect what it does? is it a `$callback`, a `$validator` ?
2. does this need to take `$arguments` to the callback at all? couldn't the callback prepare what it needs outside of this call altogether?

Reply:
same as base class, can look into that in a different PR